addons/project/models/project_project.py

```diff
@@ -1002,3 +1016,27 @@ def _add_collaborators(self, partners):
                 'partner_id': collaborator.id,
             }) for collaborator in new_collaborators],
         })
+
+    def _add_followers(self, partners):
+        self.ensure_one()
+        self.message_subscribe(partners.ids)
+
+        dict_tasks_per_partner = {}
+        dict_partners_to_subscribe_per_partner = {}
+        for task in self.task_ids:
+            tasks = dict_tasks_per_partner.get(task.partner_id, False)
+            if tasks:
+                dict_tasks_per_partner[task.partner_id] = tasks | task
+            else:
+                partners_to_subscribe = self.env['res.partner']
+                for partner in partners:
+                    if partner != task.partner_id and partner not in task.partner_id.child_ids:
+                        continue
+                    partners_to_subscribe = partners_to_subscribe | partner
+                if partners_to_subscribe:
+                    dict_tasks_per_partner[task.partner_id] = task
+                    dict_partners_to_subscribe_per_partner[task.partner_id] = partners_to_subscribe
+        for partner in dict_tasks_per_partner:
+            dict_tasks_per_partner[partner].message_subscribe(dict_partners_to_subscribe_per_partner[partner].ids)
+
+        return
```

Comment:
```suggestion
        dict_partner_ids_to_subscribe_per_partner = {}
        for task in self.task_ids:
            if task.partner_id in dict_tasks_per_partner:
                dict_tasks_per_partner[task.partner_id] |= task
            else:
                partner_ids_to_subscribe = [
                    partner.id for partner in partners
                    if partner == task.partner_id or partner in task.partner_id.child_ids
                ]
                if partner_ids_to_subscribe:
                    dict_tasks_per_partner[task.partner_id] = task
                    dict_partner_ids_to_subscribe_per_partner[task.partner_id] = partner_ids_to_subscribe
        for tasks, partner in dict_tasks_per_partner.items():
            tasks.message_subscribe(dict_partner_ids_to_subscribe_per_partner[partner])
```
- Check if partner is in `dict_tasks_per_partner` using in i/o get, because if present, it won't be empty.
- Collect the partner_ids in a list of id (so that we don't browse + anyway, we only use ids later), using a list comprehension bc why not.
- Use items() to access tasks directly.
- Delete empty return :see_no_evil: